db/db_impl_compaction_flush.cc

```diff
@@ -1541,27 +1542,49 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
       c.reset(cfd->PickCompaction(*mutable_cf_options, log_buffer));
       TEST_SYNC_POINT("DBImpl::BackgroundCompaction():AfterPickCompaction");
       if (c != nullptr) {
-        // update statistics
-        MeasureTime(stats_, NUM_FILES_IN_SINGLE_COMPACTION,
-                    c->inputs(0)->size());
-        // There are three things that can change compaction score:
-        // 1) When flush or compaction finish. This case is covered by
-        // InstallSuperVersionAndScheduleWork
-        // 2) When MutableCFOptions changes. This case is also covered by
-        // InstallSuperVersionAndScheduleWork, because this is when the new
-        // options take effect.
-        // 3) When we Pick a new compaction, we "remove" those files being
-        // compacted from the calculation, which then influences compaction
-        // score. Here we check if we need the new compaction even without the
-        // files that are currently being compacted. If we need another
-        // compaction, we might be able to execute it in parallel, so we add it
-        // to the queue and schedule a new thread.
-        if (cfd->NeedsCompaction()) {
-          // Yes, we need more compactions!
+        uint64_t size_added_by_compaction = 0;
+        // First check if we even have the space to do the compaction
+        for (size_t i = 0; i < c->num_input_levels(); i++) {
+          for (size_t j = 0; j < c->num_input_files(i); j++) {
+            FileMetaData *filemeta = c->input(i, j);
+            size_added_by_compaction += filemeta->compensated_file_size;
+          }
+        }
+        // Now figure out how much free space is in the file filesystem
+        struct statvfs statbuf;
+        statvfs(dbname_.c_str(), &statbuf);
+        uint64_t freespace = statbuf.f_bsize * statbuf.f_bfree;
+        if (freespace <= size_added_by_compaction) {
+          // Then don't do the compaction
+          c->ReleaseCompactionFiles(status);
```

Comment:
just curious, does it spam the logs as long as space is not enough? My understanding is this'll loop quickly if there's nothing else in the compaction queue.

Reply:
Right, I forgot to put a sleep here -- is that how we should handle it?

Reply:
I put a temporary sleep there, but we might want to revisit. In particular, in every test I run right now, the main problem happening isn't that compaction is too large, it's the following: Suppose max filesystem size is 10GB. compaction gets triggered, the temporary filesystem size goes up to something like 9.5GB. Compaction doesn't get a chance to clean up, and a logging thread needs to append to a log file with > 0.5GB of data (numbers are relative).
So I wonder if we should actually modify the compaction conditional to cancel compactions if compaction_size + some_buffer_room >= freespace. Where some_buffer_room is a reasonable size for log and other file appends that will be written during the duration of a compaction. For example, if all compactions are approx 3 secs, what is the amount of metadata we expect to be written those 3 secs? That amount should be some_buffer_room.

Reply:
That's a really good point. We are over-reserving a bit already because the size of the input files is typically larger than the size of the output files. That's because old versions of overwritten/deleted keys can be dropped during compaction. But yeah, estimating the amount of log+flush data during the compaction would be ideal.

Reply:
Short sleep here is ok with me, as I don't see a way to be notified of space freeing up.